### PR TITLE
Upgrade code base to Python 3

### PR DESCRIPTION
**Pull request by maximlt**
Fixes https://github.com/holoviz/param/issues/703

I ran `pyupgrade` on the code base to upgrade it to Python 3.7, and then manually made some other upgrades. Notable changes include:

- Added `watch` and `on_init` keywords to the signature of `depends`
- Deprecated `param.as_unicode`
- Deprecated `param.is_ordered_dict`

Edit:

- Added `warn` keyword to the signature of `param_union`

### CODE REVIEW

File: param/ipython.py [hunk 1]
```diff
@@ -168,7 +160,7 @@ def _build_table(self, info, order, max_col_len=40, only_changed=False):
             readonly = 'RO' if p.readonly else 'RW'
             allow_None = ' AN' if hasattr(p, 'allow_None') and p.allow_None else ''
 
-            mode = '%s %s%s' % (constant, readonly, allow_None)
+            mode = '{} {}{}'.format(constant, readonly, allow_None)
```
Comment (philippjfr) [inline param/ipython.py:163]: f-strings seem nicer.

Comment (jlstevens) [reply param/ipython.py:163]: Agreed

**Comment by jlstevens** [comment]
Looks mostly like I would expect though I would definitely prefer f-strings everywhere. Currently it is removing all the `%s` format strings and replacing it with a style that isn't that much better imho.

File: param/parameterized.py [hunk 2]
```diff
@@ -855,7 +850,7 @@ def __getattribute__(mcs,name):
 
 
 @add_metaclass(ParameterMetaclass)
```
Comment (Hoxbro) [inline param/parameterized.py:852]: `add_metaclass` is a Python 2 thing. 

https://github.com/holoviz/param/blob/2a974928537519895a814ef9c1027df7b4b6c00e/param/parameterized.py#L284-L292

Comment (maximlt) [reply param/parameterized.py:852]: Indeed, I have addressed that in another PR actually https://github.com/holoviz/param/pull/734/files#diff-c8fc4fb68012b66fbf42ad9e0b4714fd96a3a72b83558bc543130d78bd622eb3R889

**Comment by maximlt** [comment]
I've applied the suggested `ruff` command.

**Pull request merged**